trustpoint/pki/pki/request/message/rest.py

```diff
@@ -64,6 +64,9 @@ def __init__(self,
 
     
 
+    # TODO(AlexHx8472): This does not make sense I think, there can't be any exception here, the raw bytes
+    # TODO(AlexHx8472): The raw bytes should be passed and the csr should be parsed here.
+    # TODO(AlexHx8472): The serial number should not be an attribute -> access parsed data directly
```

Comment:
I agree with this, using the PKI request message framework would make more sense if bytes were passed and the CSR loaded here.
The problem is the serial number, the CSR is already parsed in `crypto_backend.py` l. 149+ in order to check that the device serial matches the CSR serial number.
It would be easiest if we changed the logic  to always use the serial from CSR, and if missing, generate it as `'tp_' + secrets.token_urlsafe(12)`.
This would mean dropping the check that CSR and manually set device serial number are identical though.
Given that the serial number is no longer a field in the device add form, this may not be an issue for initial onboarding, but I believe renewals should force the device SN to stay the same as the main persistent device entity identifier.
As far as I can see, our options would be:
- use as is, even though it does not really make sense to use PKI message without raw bytes
- remove device SN == CSR SN check, simplifying logic but potentially running into issues when renewing certificates
- add the check to `_init_csr()` (which would mix device-layer functionality into PKI layer, which I don't like)
- parsing CSR twice which is unnecessarily complex and redundant
- Doing SN check after PKI response (which is also bad, all checks should be done before issuance)
- using a separate abstraction layer within the PKI app to handle LDevID certificate issuance for onboarding, without using PKI requests. I don't like that such an implementation would likely be less portable to different CA types etc., but agree that the current rest message implementation is a bit shoehorned as the `PkiRequestMessage` framework is really meant to parse and validate the message by itself.

Reply:
I think my preferred option would be to still (optionally) pass an expected serial number to `PkiRestCsrRequestMessage` and do the validation internally, as that would streamline the logic and still make it Device layer agnostic.